blog/2023-05-11-retained-mqtt-messages-filtered.md

```diff
@@ -0,0 +1,18 @@
+---
+author: Jan Bouwhuis
+authorURL: https://twitter.com/jbouwh
+title: Recent MQTT changes to improve overall performance
+---
+
+The way Home Assistant's MQTT integration processes MQTT messages and how subscribes and unsubscribes are performed has changed.
+
+### Updates were made in the way MQTT subscribes, unsubscribes are performed
+
+Subscribes and unsubscribes are not immediately sent to the broker anymore. Debouncer code was added that will delay subscribes to the broker, and bundle them when more are added within the debounce time. At the point where the actual (un)subscribes are done, the (un)subscribes are bundled in one call to the broker. These changes will speed up the MQTT integration and the startup of integrations that register a lot of MQTT entities, especially when shared topics are subscribed (e.g. to publish the availability). In the MQTT debug logging the `mid` number shows what (un)subscribe calls have been bundled in one call to the MQTT broker.
+
+### The way retained messages are processed has changed
+
+The way how MQTT messages with a `retain` flag are handled has changed. Basically a retain flag is added to a received message when we are (re)subscribing to the broker and (some) retained payload(s) available. The broker will replay these payloads directly after (re)subscribing. Also when the connection was last and we are reconnecting the same behavior is seen. When there are multiple subscriptions that subscribe to the same topic all shared subscriptions are receiving the retained messages. Till this point the behavior does not change.
```

Comment:
Change 'The way how MQTT messages' to 'The way **that** MQTT messages'

Change 'and (some) retained payload(s)' to just 'and a retained payload' (there can be only one, and you're already saying it may or may not be there).

Change 'Also when the connection was last' to 'When the connection is the last one one that topic', and add a comma after 'reconnecting'.

Add a comma after 'to the same topic'

Change 'all shared subscriptions are receiving the retained messages' to 'all shared subscriptions receive the retained messages'

Change 'Till this point the behavior does not change.' to 'Up until this point, the behavior has not changed.'